spatialmath/mesh.go

```diff
@@ -29,10 +40,279 @@ func (m *Mesh) Triangles() []*Triangle {
+	}
+}
+
+// CollidesWith checks if the given mesh collides with the given geometry and returns true if it does.
+func (m *Mesh) CollidesWith(g Geometry, collisionBufferMM float64) (bool, error) {
+	switch other := g.(type) {
+	case *box:
+		// Mesh-ifying the box misses the case where the box encompasses a mesh triangle without its surface intersecting a triangle.
+		encompassed := m.boxIntersectsVertex(other)
+		if encompassed {
+			return true, nil
+		}
+		// Convert box to mesh and check triangle collisions
+		return m.collidesWithMesh(other.toMesh(), collisionBufferMM), nil
+	case *capsule:
+		// Use existing capsule vs mesh distance check
+		// TODO: This is inefficient! Replace with a function with a short-circuit.
+		dist := capsuleVsMeshDistance(other, m)
+		return dist <= collisionBufferMM, nil
+	case *point:
+		return m.collidesWithSphere(other.position, 0, collisionBufferMM), nil
+	case *sphere:
+		return m.collidesWithSphere(other.pose.Point(), other.radius, collisionBufferMM), nil
+	case *Mesh:
+		return m.collidesWithMesh(other, collisionBufferMM), nil
+	default:
+		return true, newCollisionTypeUnsupportedError(m, g)
+	}
+}
+
+// EncompassedBy returns whether this mesh is completely contained within another geometry.
+func (m *Mesh) EncompassedBy(g Geometry) (bool, error) {
+	if _, ok := g.(*point); ok {
+		return false, nil
 	}
+	if _, ok := g.(*Mesh); ok {
+		return false, nil
+	}
+	// For all other geometry types, check if all vertices of all triangles are inside
+	for _, pt := range m.ToPoints(1) {
+		collides, err := NewPoint(pt, "").CollidesWith(g, defaultCollisionBufferMM)
+		if err != nil {
+			return false, err
+		}
+		if !collides {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+// DistanceFrom returns the minimum distance between this mesh and another geometry.
+func (m *Mesh) DistanceFrom(g Geometry) (float64, error) {
+	switch other := g.(type) {
+	case *box:
+		// Mesh-ifying the box misses the case where the box encompasses a mesh triangle without its surface intersecting a triangle.
+		encompassed := m.boxIntersectsVertex(other)
+		if encompassed {
+			return 0, nil
+		}
+		return m.distanceFromMesh(other.toMesh()), nil
+	case *capsule:
+		return capsuleVsMeshDistance(other, m), nil
+	case *point:
+		return m.distanceFromSphere(other.position, 0), nil
+	case *sphere:
+		return m.distanceFromSphere(other.pose.Point(), other.radius), nil
+	case *Mesh:
+		return m.distanceFromMesh(other), nil
+	default:
+		return math.Inf(-1), newCollisionTypeUnsupportedError(m, g)
+	}
+}
+
+// Returns true if any triangle vertex of the mesh intersects the box.
+func (m *Mesh) boxIntersectsVertex(b *box) bool {
+	for _, p := range m.ToPoints(1) {
+		if pointVsBoxCollision(p, b, defaultCollisionBufferMM) {
+			return true
+		}
+	}
+	return false
+}
+
+func (m *Mesh) distanceFromSphere(pt r3.Vector, radius float64) float64 {
+	minDist := math.Inf(1)
+
+	for _, tri := range m.triangles {
+		worldTri := NewTriangle(
```

Comment:
I think its probably worth defining a transform function on a triangle to avoid a bunch of repeated code here

Reply:
I'm hesitant to do so. This is one of those cases where that would actually have performance implications.
The meshes we process are beginning to have large numbers of triangles, and that number continues to grow. Per-triangle function calls become expensive due to the simple overhead of the function call. Also it removes the ability to short-circuit early on collision checks.